### PR TITLE
checking of constant/writable arguments

### PR DESCRIPTION
**Pull request by bgregoir**

### CODE REVIEW

File: compiler/src/pretyping.ml [hunk 1]
```diff
@@ -1217,7 +1217,7 @@ let tt_lvalue pd (env : 'asm Env.env) { L.pl_desc = pl; L.pl_loc = loc; } =
 
   | S.PLVar x ->
     let x = tt_var `NoParam env x in
-    reject_constant_pointers loc x ;
+(*    reject_constant_pointers loc x ; *)
```
Comment (vbgl) [inline compiler/src/pretyping.ml:1220]: Maybe this line should be removed completely.

**Comment by eponier** [comment]
This example is artificial, but do we want to accept such things?

```
fn f (reg ptr u64[1] r) -> reg u64 {
  reg u64 res;
  stack u64[1] s;

  r = s;
  r[0] = 1;
  res = r[0];

  return res;
}

export fn main () -> reg u64 {
  reg u64 res;
  stack u64[1] s;

  s[0] = 1;

  res = f(s);

  return res;
}
```

**Review by vbgl — CHANGES_REQUESTED** [review]
This is not ready. The following program is now rejected:

~~~
fn zerofill(reg mut ptr u64[1] x) -> reg ptr u64[1] {
  x[0] = 0;
  return x;
}

export
fn main() -> reg u64 {
  stack u64[1] s;
  s = zerofill(s);
  reg u64 r;
  r = s[0];
  return r;
}
~~~

**Comment by eponier** [comment]
With @bgregoir, we are not sure what to do exactly, so let's postpone this a bit.

**Comment by vbgl** [comment]
IMHO, it is OK to reject the artificial example above.

**Comment by bgregoir** [comment]
I did not see your last commit and I did the same patch. 
Anyway I think we try to ensure something that is verified later by the compiler without having all the information.
We can do a variation of zerofill example:

inline fn zerofill(stack u64[1] x) -> stack u64[1] {
  x[0] = 0;
  return x;
}

export
fn main() -> reg u64 {
  stack u64[1] s;
  reg const ptr u64[1] t;
  t = s;
  s = zerofill(t);
  reg u64 r;
  r = s[0];
  return r;
}

File: compiler/src/pretyping.ml [hunk 2]
```diff
@@ -1595,18 +1594,39 @@ let check_lval_pointer loc x =
   | _ -> rs_tyerror ~loc (NotAPointer x)
 
 let mk_call loc is_inline lvs f es =
-  begin match f.P.f_cc with 
-  | P.Internal | P.Export -> ()
-  | P.Subroutine _ ->
+  let open P in
+  begin match f.f_cc with
+  | Internal -> ()
+  | Export -> if is_inline = E.InlineFun then rs_tyerror ~loc (string_error "call to export function needs to be inlined")
```
Comment (vbgl) [inline compiler/src/pretyping.ml:1600]: This test is wrong…

Comment (vbgl) [reply compiler/src/pretyping.ml]: Fixed!

**Comment by vbgl** [comment]
Unless someone wants to implement a more precise check, I recommend to merge this PR as is.

**Comment by bgregoir** [comment]
Ok with this.

**Comment by bgregoir** [comment]
Apparently I can't merge it myself.

**Comment by eponier** [comment]
You have to accept it first (i.e. add a positive review).

**Review by eponier — APPROVED** [review]
@bgregoir accepts his own PR

**Pull request merged**